From 8f1a48d6b8168b5f6bd77cc4c60adddf23e77814 Mon Sep 17 00:00:00 2001
From: Mohammed Rafi KC <rkavunga@redhat.com>
Date: Tue, 8 Dec 2015 19:55:22 +0530
Subject: [PATCH 177/178] tier/dht: files are still going to decommissioned subvol

After detach tier start, creates are still going to hot
tier. Because when creating data files we are not checking for
decommissioned bricks.
Backpor of>
>Change-Id: I8e28258d9b2367dcc8ad6e5e91d0e54d92fdf771
>BUG: 1289602
>Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>
>Reviewed-on: http://review.gluster.org/12914
>Tested-by: Gluster Build System <jenkins@build.gluster.com>
>Reviewed-by: Dan Lambright <dlambrig@redhat.com>
>Tested-by: Dan Lambright <dlambrig@redhat.com>

>(cherry picked from commit cb41de7c4a7d620dbab645f2b105286b68bde9a9)
>Change-Id: I56e4551657a234ac51fa2513dc46a48c19c4fac7
>BUG: 1289602
>Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>

Change-Id: I1745a49f2b5fdb4f9285ed4356e7c5d4a35ecf18
BUG: 1287532
Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/63462
Reviewed-by: Joseph Fernandes <josferna@redhat.com>
Tested-by: Joseph Fernandes <josferna@redhat.com>
---
 xlators/cluster/dht/src/dht-common.c |   30 +++++++++++++++++++++++++++---
 1 files changed, 27 insertions(+), 3 deletions(-)

diff --git a/xlators/cluster/dht/src/dht-common.c b/xlators/cluster/dht/src/dht-common.c
index 6a71767..bc7390b 100644
--- a/xlators/cluster/dht/src/dht-common.c
+++ b/xlators/cluster/dht/src/dht-common.c
@@ -6038,6 +6038,27 @@ err:
         return 0;
 }
 
+gf_boolean_t
+dht_is_hot_tier_decommissioned (xlator_t *this)
+{
+        dht_conf_t              *conf        = NULL;
+        xlator_t                *hot_tier    = NULL;
+        int                      i           = 0;
+
+        conf = this->private;
+        hot_tier = conf->subvolumes[1];
+
+        if (conf->decommission_subvols_cnt) {
+                for (i = 0; i < conf->subvolume_cnt; i++) {
+                        if (conf->decommissioned_bricks[i] &&
+                                conf->decommissioned_bricks[i] == hot_tier)
+                                return _gf_true;
+                }
+        }
+
+        return _gf_false;
+}
+
 int
 dht_create_tier_wind_to_avail_subvol (call_frame_t *frame, xlator_t *this,
                                       xlator_t *subvol, loc_t *loc, int32_t flags,
@@ -6058,9 +6079,12 @@ dht_create_tier_wind_to_avail_subvol (call_frame_t *frame, xlator_t *this,
         if (conf->subvolumes[0] != cold_subvol) {
                 hot_subvol = conf->subvolumes[0];
         }
-
-        /* if hot tier full, write to cold */
-        if (dht_is_subvol_filled (this, hot_subvol)) {
+        /*
+         * if hot tier full, write to cold.
+         * Also if hot tier is full, create in cold
+         */
+        if (dht_is_subvol_filled (this, hot_subvol) ||
+            dht_is_hot_tier_decommissioned (this)) {
                 gf_msg_debug (this->name, 0,
                               "creating %s on %s", loc->path,
                               cold_subvol->name);
-- 
1.7.1

